### PR TITLE
TOOL-8 - Implement Router Service

### PR DESCRIPTION
**Pull request by ann-sargsyan**

### CODE REVIEW

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java [hunk 1]
```diff
@@ -12,45 +13,73 @@
...
+                                                          @RequestParam(value = "zipCodes", required = false) List<String> zipCodes,
+                                                          @RequestParam(value = "coordinates", required = false) List<String> coordinates,
+                                                          @RequestParam(value = "serviceApi", defaultValue = "OPEN_WEATHER_MAP", required = false) ServiceApiEnum inputOfServiceApi
     ) {
-        return getWeatherResponse(cities, inputOfServiceApi);
+
+        List<CoordinateWeatherRequest> coordinateRequests = createCoordinateRequests(coordinates);
+        List<ZipCodeWeatherRequest> zipCodeWeatherRequests = createZipCodeRequest(zipCodes);
+
+        WeatherRequest weatherRequest = WeatherRequest.builder().cities(cities).zipcode(zipCodeWeatherRequests).coordinates(coordinateRequests).service(inputOfServiceApi).build();
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: let's build a request from the parameters that come in, process it inside the services, or in a separate class, not in the controller

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java [hunk 2]
```diff
@@ -12,45 +13,73 @@
...
-        OpenWeatherRequest weatherRequest = OpenWeatherRequest.builder()
-                .cities(cities)
-                .service(serviceApi)
-                .build();
-        log.info(REQUEST_MESSAGE, weatherRequest.service());
-        WeatherResponse weatherResponse = openWeatherService.getWeather(weatherRequest);
-        return ResponseEntity.ok(weatherResponse);
+    private List<CoordinateWeatherRequest> createCoordinateRequests(List<String> coordinates) {
+        return Optional.ofNullable(coordinates)
+                .map(coords -> IntStream.range(ZERO, coords.size() / TWO)
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: let's accept coordinates in the line "1.00:2.00" so it will be easier to validate

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java [hunk 3]
```diff
@@ -12,45 +13,73 @@
...
 
-    private ResponseEntity<WeatherResponse> getWeatherResponse(List<String> cities, ServiceApiEnum serviceApi) {
-        OpenWeatherRequest weatherRequest = OpenWeatherRequest.builder()
-                .cities(cities)
-                .service(serviceApi)
-                .build();
-        log.info(REQUEST_MESSAGE, weatherRequest.service());
-        WeatherResponse weatherResponse = openWeatherService.getWeather(weatherRequest);
-        return ResponseEntity.ok(weatherResponse);
+    private List<CoordinateWeatherRequest> createCoordinateRequests(List<String> coordinates) {
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: it is necessary to transfer the processing of this data inside the service

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java [hunk 4]
```diff
@@ -12,45 +13,73 @@
...
+                .map(coords -> IntStream.range(ZERO, coords.size() / TWO)
+                        .mapToObj(i -> CoordinateWeatherRequest.builder()
+                                .lat(Double.valueOf(coords.get(i * TWO)))
+                                .lon(Double.valueOf(coords.get(i * TWO + ONE)))
+                                .build())
+                        .collect(Collectors.toList()))
+                .orElse(new ArrayList<>());
+    }
+
+    private List<ZipCodeWeatherRequest> createZipCodeRequest(List<String> zipCodes) {
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: it is necessary to transfer the processing of this data inside the service

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java [hunk 5]
```diff
@@ -12,45 +13,73 @@
...
+        return Optional.ofNullable(zipCodes)
+                .map(zips ->
+                        IntStream.range(ZERO, zips.size() / TWO)
+                                .mapToObj(i ->
+                                        ZipCodeWeatherRequest.builder()
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: Also. I would suggest putting object construction into a separate method

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java [hunk 6]
```diff
@@ -12,45 +13,73 @@
...
-                .cities(cities)
-                .service(serviceApi)
-                .build();
-        log.info(REQUEST_MESSAGE, weatherRequest.service());
-        WeatherResponse weatherResponse = openWeatherService.getWeather(weatherRequest);
-        return ResponseEntity.ok(weatherResponse);
+    private List<CoordinateWeatherRequest> createCoordinateRequests(List<String> coordinates) {
+        return Optional.ofNullable(coordinates)
+                .map(coords -> IntStream.range(ZERO, coords.size() / TWO)
+                        .mapToObj(i -> CoordinateWeatherRequest.builder()
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: Also. I would suggest putting object construction into a separate method

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java [hunk 7]
```diff
@@ -12,45 +13,73 @@
...
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 
 @RequiredArgsConstructor
 @RestController
 @RequestMapping("/weather")
 @Slf4j
 public class WeatherController {
+    private static final Integer ZERO = 0;
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: there is no great need to put such simple variables into constants, this complicates the code and makes it less readable, we must try to maintain balance

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java [hunk 8]
```diff
@@ -12,45 +13,73 @@
...
     ) {
-        return getWeatherResponse(cities, inputOfServiceApi);
+
+        List<CoordinateWeatherRequest> coordinateRequests = createCoordinateRequests(coordinates);
+        List<ZipCodeWeatherRequest> zipCodeWeatherRequests = createZipCodeRequest(zipCodes);
+
+        WeatherRequest weatherRequest = WeatherRequest.builder().cities(cities).zipcode(zipCodeWeatherRequests).coordinates(coordinateRequests).service(inputOfServiceApi).build();
+        log.info(REQUEST_MESSAGE, weatherRequest);
+        WeatherService weatherService = weatherRouterService.getWeatherBySelectedService(inputOfServiceApi);
+        WeatherResponse weatherResponse = weatherService.getWeather(weatherRequest);
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]:
```suggestion
        WeatherResponse weatherResponse = weatherRouterService.getWeatherBySelectedService(inputOfServiceApi).getWeather(weatherRequest);
```
Having a record in 2 lines does not give us any advantages

File: src/main/java/com/example/weatherwaveapi/service/WeatherRouterService.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class WeatherRouterService {
+    private static final String NOT_SUPPORTED_MESSAGE = "Not supported service";
+    private final OpenWeatherService openWeatherService;
+    private final YandexWeatherService yandexWeatherService;
+
+
+    public WeatherService getWeatherBySelectedService(ServiceApiEnum service) {
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/WeatherRouterService.java]: please rename it to getWeatherService

File: src/main/java/com/example/weatherwaveapi/service/WeatherRouterService.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        return switch (service) {
+            case OPEN_WEATHER_MAP -> openWeatherService;
+            case YANDEX -> yandexWeatherService;
+            default -> throw new IllegalArgumentException(NOT_SUPPORTED_MESSAGE);
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/WeatherRouterService.java]: There is no point in issuing such a simple error message, it also complicates the code and can be misleading

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java (hunk 1 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: I get it, I will do it

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java (hunk 2 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: I changed it

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java (hunk 3 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: Ok, I did it

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java (hunk 4 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: I did it

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java (hunk 5 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: I have done it

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java (hunk 6 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: I did it

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java (hunk 7 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: I have removed it

File: src/main/java/com/example/weatherwaveapi/controller/WeatherController.java (hunk 8 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/controller/WeatherController.java]: I did it

File: src/main/java/com/example/weatherwaveapi/service/WeatherRouterService.java (hunk 10 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/service/WeatherRouterService.java]: Ok, I did it

File: src/main/java/com/example/weatherwaveapi/service/WeatherRouterService.java (hunk 9 above)
Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/service/WeatherRouterService.java]: I changed it

File: src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java [hunk 11]
```diff
@@ -78,4 +83,22 @@ private YandexApiContainer getYandexApiContainer(String url, String format) {
...
+    private List<CoordinateWeatherRequest> convertCoordinatesRequest(List<String> coordinates) {
+        return Optional.ofNullable(coordinates)
+                .map(this::coordinateBuilder)
+                .orElse(new ArrayList<>());
+    }
+
+    private List<CoordinateWeatherRequest> coordinateBuilder(List<String> coordinates) {
+        return coordinates.stream()
+                .map(coordinate -> coordinate.split(":"))
+                .filter(splitCoordinate -> splitCoordinate.length == 2)
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java]: what happens if there are not 2 coordinates? do we just ignore this situation? we need to handle all possible situations: Check if the list is empty, the record format is invalid, the data format is incorrect. in other words, we need to validate the data and report errors to the user, if it concerns him, so that he can correct his request

File: src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java [hunk 12]
```diff
@@ -78,4 +83,22 @@ private YandexApiContainer getYandexApiContainer(String url, String format) {
...
+                .map(splitCoordinate ->
+                        CoordinateWeatherRequest.builder()
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java]: please move the builder into a separate method

File: src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java [hunk 13]
```diff
@@ -78,4 +83,22 @@ private YandexApiContainer getYandexApiContainer(String url, String format) {
...
+                                .lat(Double.parseDouble(splitCoordinate[0]))
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java]: what happens if the data cannot be converted to a double type?

File: src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java [hunk 14]
```diff
@@ -137,4 +139,21 @@ private void validateZipCode(Integer zipcode) {
...
     }
+
+    private List<ZipCodeWeatherRequest> convertZipCodeRequest(List<String> zipCodes) {
+        return Optional.ofNullable(zipCodes)
+                .map(this::zipCodeBuilder)
+                .orElse(new ArrayList<>());
+    }
+
+    private List<ZipCodeWeatherRequest> zipCodeBuilder(List<String> zipCodes) {
+        return IntStream.range(0, zipCodes.size() / 2)
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java]: Would you like to make the same changes here as with the coordinates?

File: src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java [hunk 15]
```diff
@@ -116,7 +119,6 @@ private WeatherOpenApiContainer getForecastContainerByCity(String city) {
     private WeatherApiResponse convertContainer(WeatherOpenApiContainer container) {
         return Optional.ofNullable(container.errorMessage())
                 .map(error -> WeatherApiResponse.builder()
-                        .success(false)
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java]: why was this line removed?

Comment (mkebets) [reply src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java]: let's leave in all replies a field with a flag indicating whether everything went well or not

Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java]: Last time you said me to remove it because in json we don't have this field

Comment (mkebets) [reply src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java]: let's leave it anyway.

File: src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java [hunk 16]
```diff
@@ -137,4 +140,36 @@ private void validateZipCode(Integer zipcode) {
             throw new InvalidZipCodeException(INVALID_ZIPCODE_EXCEPTION_MESSAGE);
         }
     }
+
+    private List<ZipCodeWeatherRequest> convertZipCodeRequest(List<String> zipCodes) {
+        if (zipCodes == null || zipCodes.isEmpty() || zipCodes.size() % 2 != 0) {
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java]:
```suggestion
        if (CollectionUtils.isEmpty(zipCodes) || zipCodes.size() % 2 != 0) {
```

File: src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java [hunk 17]
```diff
@@ -137,4 +140,36 @@ private void validateZipCode(Integer zipcode) {
...
+            System.err.println("Invalid zip code list");
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/OpenWeatherService.java]:
```suggestion
           log.info("invalid zip code format");
```

File: src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java [hunk 18]
```diff
@@ -73,9 +77,49 @@ private YandexApiContainer getYandexApiContainer(String url, String format) {
...
+        coordinates.forEach(coordinate -> {
+            String[] splitCoordinate = coordinate.split(":");
+
+            if (splitCoordinate.length == 2) {
+                try {
+                    double lat = Double.parseDouble(splitCoordinate[0]);
+                    double lon = Double.parseDouble(splitCoordinate[1]);
+                    result.add(buildCoordinateRequest(lat, lon));
+                } catch (NumberFormatException e) {
+                    errors.add("Invalid coordinate format for: " + coordinate);
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java]:
```suggestion
                    log.error("Invalid coordinate format for: {}", coordinate);
```

File: src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java [hunk 19]
```diff
@@ -73,9 +77,49 @@ private YandexApiContainer getYandexApiContainer(String url, String format) {
...
+        if (coordinates == null || coordinates.isEmpty()) {
+            System.err.println("Invalid coordinates list");
+            return Collections.emptyList();
+        }
+        return coordinateBuilder(coordinates);
+    }
+
+    private List<CoordinateWeatherRequest> coordinateBuilder(List<String> coordinates) {
+        List<CoordinateWeatherRequest> result = new ArrayList<>();
+        List<String> errors = new ArrayList<>();
```
Comment (mkebets) [inline src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java]: this is an overkill solution, just let's log errors as they appear

Comment (ann-sargsyan) [reply src/main/java/com/example/weatherwaveapi/service/YandexWeatherService.java:96]: You mean just change errors printing to logging?

**Review by mkebets — APPROVED** [review]
